support hmis okta sign on

ttoomey opened the pull request:
### _Please squash-merge this PR_

Reference
[#183560765](https://www.pivotaltracker.com/story/show/183560765)
[hmis frontend #195](https://github.com/greenriver/hmis-frontend/pull/195)

Adjust okta integration to support login from the HMIS front end. The expectation is that the HMIS front end will use a distinct okta application.

The provider fields on user model are deprecated in favor of `oauth_identities` to allow a user to authenticate with more than one provider.

An "active_provider" cookie is stored so we know how to redirect the user after logout.

Per spec, added a settings endpoint `Hmis::AppSettingsController` for the HMIS front-end. For now this is just the okta endpoint, if enabled.

**testing:**
Tested against okta dev server. Requires creating a new okta "app":
```
Sign-in redirect URIs: https://hmis.dev.test/hmis/users/auth/okta/callback
Sign-out redirect URIs: https://hmis.dev.test
```
Then the client id and key should be set as env variables
```
HMIS_OKTA_CLIENT_ID=id_here
HMIS_OKTA_CLIENT_SECRET=secret_here
```

**misc:**
* use bang variant (`save!`) in in omniauth methods so save-failures are not silently ignored
* tidy logging in omniauth callback controller

ttoomey committed:
Support okta SSO for HMIS 183560765

ttoomey committed:
Merge branch 'stable' into 183560765-okta-hmis-support_2

ttoomey committed:
gem updates for CI

eanders committed:
Merge branch 'release-68' into 183560765-okta-hmis-support_2

eanders commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -38,7 +40,62 @@ def passthru
+    cookies['CSRF-Token'] = form_authenticity_token
+  end
+
+  USER_TYPES = [HMIS_USER_TYPE = 'hmis_user'.freeze, WH_USER_TYPE = 'user'.freeze].freeze
```
Since these are used above, can we move them to the top of the file?

eanders commented on config/initializers/oauth2.rb:
```diff
@@ -0,0 +1,10 @@
+
+OAuth2.configure do |config|
+  # https://gitlab.com/oauth-xx/oauth2#global-configuration
+  config.silence_extra_tokens_warning = true # default: false
```
Thank you!

eanders commented on db/seeds.rb:
```diff
@@ -332,7 +332,9 @@ def setup_hmis_admin_access
 
   user.hmis_data_source_id = hmis_ds.id
-  user.user_hmis_data_sources_roles.create(role: role, data_source_id: user.hmis_data_source_id)
+  access_group = Hmis::AccessGroup.create!(name: 'test')
```
I don't _think_ we want to do this here without some guard on when it gets called.  `db:seed` is called on every deploy, and we don't want a test access group installed for production.
If this is used for testing, it could just get wrapped in an `if Rails.env.test?` or move to a factory.

ttoomey replied:
@eanders - Thanks, I didn't realize we were running this on deploy. I added some documentation for this to the top of the seeds file.

I noticed that `user.user_hmis_data_sources_roles` does not exist. Perhaps this line should just be removed?

eanders replied:
@gigxz any thoughts on `user.user_hmis_data_sources_roles`, I'm not familiar with that one.  Maybe that's obsolete after @ZachCMP 's recent upgrade to ACLs?  

Thanks @ttoomey, I think this looks good.

ttoomey replied:
thanks @eanders - my understanding is that I should rebase this into r70 once the sprockets issue is resolved.

eanders replied:
@ttoomey you can give https://github.com/greenriver/hmis-warehouse/pull/2910 a try and see if that solves it for you.  (It should.)  Otherwise, it's totally fine to wait until that gets merged into `release-70`

gigxz replied:
Yeah, I have some code in my local tasks for seeding the HMIS admin roles. I'll go back and update this file to be correct. We shouldn't be auto-assigning HMIS admin perms to a random user, so I'll remove that part.

ttoomey committed:
address feedback 183560765

ttoomey committed:
Merge branch '183560765-okta-hmis-support_2' of github.com:greenriver/hmis-warehouse into 183560765-okta-hmis-support_2

gigxz commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -6,27 +6,30 @@
   include AuthenticatesWithTwoFactor
+  USER_TYPES = [HMIS_USER_TYPE = 'hmis_user'.freeze, WH_USER_TYPE = 'user'.freeze].freeze
 
   def okta
```
If we're calling this endpoint from the HMIS, we also need to add it to the rack cors config https://github.com/greenriver/hmis-warehouse/blob/release-70/config/initializers/cors.rb. Currently, we only allow requests to `/hmis/*` endpoints from the HMIS. I think it is safest to call out the specific endpoints, so `/users/auth/okta` instead of `/users/auth`.

For the reverse-proxying to work, you'll also need to make a copy of this block https://github.com/greenriver/hmis-frontend/blob/13f0af85c014f813e0d17baa898bb72c029aadf4/docker/templates/default.conf.template#L26-L36 for the new locations. Again, I think it's probably safest to proxy only the specific routes that we need.

ttoomey replied:
Thanks @gigxz! I've added the auth end point to the cors config. Good point about making the the endpoint more specific. I'll adjust the the nginx config template for the hmis-frontend too.

ttoomey committed:
add cors for HMIS okta 183560765

ttoomey committed:
Merge branch 'release-70' into 183560765-okta-hmis-support_2

ttoomey commented:
keeping as draft; will likely need to support multiple okta apps

ttoomey committed:
hmis app settings endpoint

ttoomey committed:
support more than one okta app, distinct endpoints

- raise if hmis user requires 2fa
- support multiple oauth "identities"
- support auth for more than one model
- replace devise's omniauthable
- extract user creation into service object

ttoomey committed:
Merge branch 'release-70' into 183560765-okta-hmis-support_2

ttoomey committed:
undo accidental schema, cleanup

ttoomey commented on config/initializers/omniauth.rb:
```diff
@@ -0,0 +1,74 @@
+    end
+  end
+
+  OmniAuth.config.on_failure = Proc.new { |env|
```
TBD- need to find a way to test failure condition

ttoomey committed:
bugfix multi-auth providers for user admin

ttoomey committed:
fix for multi auth-providers

ttoomey committed:
fixup endpoints

ttoomey committed:
fixup omniauth failure handler

eanders commented on app/controllers/concerns/omniauth_callback_behavior.rb:
```diff
@@ -0,0 +1,61 @@
+  protected
+
+  def home_path
+    root_path
```
This should be `user.my_root_path` for warehouse logins.

ttoomey replied:
Looking at it again, this was only used by passthru, which was a vestige of devis/omniauthable. I don't think it's needed, removed this method

eanders commented on app/helpers/application_helper.rb:
```diff
@@ -389,4 +389,12 @@ def hmis_enabled?
   end
+
+  def omni_auth_providers
+    if ENV['ENABLE_HMIS_API']
+      [['Okta', '/users/auth/okta']]
+    else
```
I'm probably missing something, but not _all_ HMIS installations will have an OKTA provider, only one at the moment, so that should probably be pulling from a database table or ENV.

ttoomey replied:
Thanks for spotting that, looks like I fat-fingered the wrong env variable there, my intention was `if ENV['OKTA_DOMAIN']`

This is is a replacement for  `resource_class.omniauth_providers` in the warehouse UI. If we need to move the provider config to the db we should check in on that since I've probably missed part of the spec

eanders replied:
Nope, `OKTA_DOMAIN` makes way more sense to me.

eanders commented on app/models/omniauth_user_provisioner.rb:
```diff
@@ -0,0 +1,75 @@
+    OauthIdentity.transaction do
+      attrs = { provider: auth.provider, uid: auth.uid }
+      # create an orphan authorzation to lock before creating a user
+      now = Time.now
```
```suggestion
      now = Time.current
```

ttoomey committed:
PR feedback, bugfixes

ttoomey committed:
restore devise failure behavior, support per-provider failure

ttoomey committed:
fix for empty string in env var (broke tests)

ttoomey commented:
Note, test failures seem to be related to `TodoOrDie::OverdueTodo` - I think they are unrelated to this PR

ttoomey committed:
small cleanups

eanders reviewed (approved):
Looks good to me!  We should confirm on staging that it doesn't break the warehouse logins before merging.

gigxz commented:
This and frontend look good to me! It did appear the base branches were somewhat out of sync - frontend was expecting a graphql attribute that wasn't present yet (2 prs merged by zach last week). Before deploying, can you merge in the latest release-70 to each?

ttoomey committed:
Merge branch 'release-70' into 183560765-okta-hmis-support_2

ttoomey committed:
undo schema changes

ttoomey committed:
bugfix in okta notification

ttoomey committed:
report oauth failures to sentry

ttoomey committed:
additional okta debugging to sentry

ttoomey committed:
enable origin forgery protection, logging

ttoomey committed:
test to ensure user with provider cannot login

eanders commented on spec/requests/sessions_controller_spec.rb:
```diff
@@ -20,6 +20,10 @@
       post user_session_path(user: { email: user.email, password: 'incorrect' })
     end
 
+    it 'user sees and error' do
```
```suggestion
    it 'user sees an error' do
```

ttoomey committed:
Update spec/requests/sessions_controller_spec.rb

Co-authored-by: Elliot <elliot@greenriver.org>

ttoomey committed:
fix okta test, the test env has okta disabled

eanders commented on spec/requests/sessions_controller_spec.rb:
```diff
@@ -141,4 +145,16 @@
       end
     end
   end
+
```
@ttoomey is this to work in conjunction with https://github.com/greenriver/hmis-warehouse/blob/e089fb53315c0b97bfb6a493231fd20115ad418e/.github/workflows/rails_tests.yml#L181-L183 or should we do something different/move this to a different test?  Maybe the old tests aren't relevant?

ttoomey replied:
I was just running the test locally where I can manage the env. I didn't realize we had  this one-off in CI already, that makes sense. Should I just add this test?

```diff
--- a/.github/workflows/rails_tests.yml
+++ b/.github/workflows/rails_tests.yml
@@ -180,7 +180,7 @@ jobs:
           LOG_LEVEL: INFO
         run: |
           bundle exec rspec --fail-fast=$MAX_FAILURES --color --format p -P "{spec,drivers}/**/*_spec.rb"
-          OKTA_DOMAIN=localhost OKTA_CLIENT_ID=x OKTA_CLIENT_SECRET=x bundle exec rspec --fail-fast=$MAX_FAILURES --color -fd spec/requests/omniauth_spec.rb # this one test needs a different ENV
+          OKTA_DOMAIN=localhost OKTA_CLIENT_ID=x OKTA_CLIENT_SECRET=x bundle exec rspec --fail-fast=$MAX_FAILURES --color -fd spec/requests/omniauth_spec.rb spec/requests/sessions_controller_spec.rb # this one test needs a different ENV
       - name: Run logging tests
```

II think the tests in session_conroller_spec and omniauth_spec are still relevant

eanders replied:
I think so.  Thanks!

ttoomey committed:
add one-off okta test for CI

ttoomey committed:
fix test for empty string in env variable

ttoomey committed:
Merge branch 'release-70' into 183560765-okta-hmis-support_2

gigxz commented on config/environments/development.rb:
```diff
@@ -127,9 +127,4 @@
   routes.default_url_options ||= {}
   routes.default_url_options[:script_name]= ''
-
-  # Disable CSRF origin check during development because the HMIS frontend has a different origin
```
This is still needed for those of us who are developing the HMIS locally @ttoomey. Maybe you didn't need it because of your docker setup? Anyway, I'm adding it back

ttoomey replied:
Sorry about that! I took it out to reproduce an issue I saw on QA related to rails origin forgery protection.

gigxz replied:
Resolved with https://github.com/greenriver/hmis-frontend/pull/212 which makes [this function](https://apidock.com/rails/v5.1.7/ActionController/RequestForgeryProtection/valid_request_origin%3F) equally happy.
